Add delay message

NickGreen opened the pull request:
Adds delay message to the update message box for each eligible plugin on the plugins page.
Displays date in format as set in the site settings, in UTC.

<img width="1221" alt="Screenshot 2024-10-10 at 12 39 40" src="https://github.com/user-attachments/assets/43e4a718-515e-41c9-a6e3-760b5e2e0531">

NickGreen committed:
Add delay message to update box on plugins page

NickGreen committed:
PHPCS

georgestephanis commented:
I'm not certain on the wording -- I think my brain would prefer the append as something like `Version ##.## will not autoupdate until at least MM/DD/YY` or the like -- I appreciate how we're using the site's options for date and time format -- but I'm largely going to defer to TAMs for how they feel the wording would work best for partners.

NickGreen committed:
Prevent null warnings with null array key

NickGreen committed:
Add delay message even if not during business hours

NickGreen commented:
Updated the wording as per TAM suggestion.

@georgestephanis @ecairol Would either of you dudes be willing to take a look at the code to make sure it's right and safe, before we merge it and send it off to a bunch of sites? :)

ecairol commented on class-plugin-autoupdate-filter.php:
```diff
@@ -146,24 +146,47 @@ public function filter_maybe_disable_all_autoupdates( $update, $item ): bool {
+				add_filter(
+					"in_plugin_update_message-{$plugin_file}",
+					function() use ( $plugin_new_version, $formatted_date ) {
+						echo ' For stability, autoupdates operate on a slight delay. Autoupdate to version ' . esc_html( $plugin_new_version ) . ' is currently estimated to run after ' . esc_html( $formatted_date ) . ' UTC.';
```
Here, `$formatted_date` mentions UTC, but it was formatted using a localized function. Does `date_i18n()` change the date to the site's timezone?

NickGreen replied:
Good catch! It will actually be displaying in the timezone of the WordPress site itself. So I guess the question is: do we want it to display in UTC?

If so, we can just set the gmt parameter to true like this:

```PHP
// Set the $gmt parameter to true for UTC time
$formatted_date = date_i18n( $datetime_format, $delay_date, true );
```

NickGreen replied:
I'm not sure which way is better.

ecairol replied:
Given the number of sites this will be displayed, UTC may be better.

NickGreen replied:
Fixed. It now displays in UTC.

ecairol commented on class-plugin-autoupdate-filter.php:
```diff
@@ -146,24 +146,47 @@ public function filter_maybe_disable_all_autoupdates( $update, $item ): bool {
+			$option_key = 'plugin_update_delays';
+			$delays     = get_option( $option_key, array() );
+			if ( isset( $delays[ $plugin_file ][ $plugin_new_version ] ) ) {
+				$delay_date = $delays[ $plugin_file ][ $plugin_new_version ];
```
I think it's worth checking `is_numeric( $delay_date )` before using it on `date_i18n()`

NickGreen replied:
Thanks! Added.

ecairol reviewed (commented):
LGTM! Left a couple of comments.

NickGreen committed:
Set to UTC time

NickGreen committed:
Check if timestamp is numeric to prevent warnings

NickGreen committed:
Don't show msg if autoupdates aren't available or if plugin is deactivated

e.g. If not connected to WC.com, updates for paid plugins won't be available, so we shouldn't show the autoupdated message. Also, plugins won't autoupdate if deactivated, so no message.

NickGreen commented:
Also added checks to not show the message if:
1. Autoupdates aren't available (i.e. wc.com isn't connected, so paid plugins can't update)
2. Plugin is deactivated (autoupdates don't happen when plugins aren't active)

Tested, merging.
